### PR TITLE
Boost: Add endpoint to list source providers

### PR DESCRIPTION
**Pull request by dilirity**
Created for https://github.com/Automattic/boost-cloud/pull/432

It would also allow us to develop a tool that generates critical CSS, as the source providers (groups) are no longer private to the website.

## Proposed changes:

* Add endpoint to list source providers.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

n/a

## Does this pull request change what data or activity we track or use?

no

## Testing instructions:

* Make sure the endpoint is visible after applying the patch.

### CODE REVIEW

File: projects/plugins/boost/app/rest-api/endpoints/List_Source_Providers.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	public function response( $_request ) {
+		$providers = new Source_Providers();
+		return rest_ensure_response( $providers->get_provider_sources() );
+	}
+
+	public function permissions() {
+		return array();
+	}
```
Comment (thingalon) [inline projects/plugins/boost/app/rest-api/endpoints/List_Source_Providers.php]: If we set `permissions` to `Signed_With_Blog_Token` ([example](https://github.com/Automattic/jetpack/blob/f98e248acf2ef55dee12a5558df9ab1401c638a0/projects/plugins/boost/app/rest-api/endpoints/Update_Cloud_CSS.php#L113)) then we can only allow access from WordPress.com.

**Review by thingalon — CHANGES_REQUESTED** [review]
I think this is a good idea.

However, I don't think we should add this as an insecure API endpoint. What if there are custom post types which are supposed to be private? We could leak information.

I've suggested in-line how we can/should lock this down to be callable only from WordPress.com

File: projects/plugins/boost/app/rest-api/endpoints/List_Source_Providers.php (hunk 1 above)
Comment (dilirity) [reply projects/plugins/boost/app/rest-api/endpoints/List_Source_Providers.php]: Yep yep yep! I agree with this. Added in 366bc2a.

**Comment by dilirity** [comment]
The format we currently use is this:

```
[
	{
		"key": "core_front_page",
		"label": "Front page",
		"urls": ["http://jetpack-boost.test/"],
		"success_ratio": 1
	},
	{
		"key": "core_posts_page",
		"label": "Posts page",
		"urls": ["http://jetpack-boost.test/"],
		"success_ratio": 1
	},
	{
		"key": "singular_page",
		"label": "Single page view",
		"urls": [
			"http://jetpack-boost.test/another-parent/this-is-a-somewhat-long-url/",
			"http://jetpack-boost.test/another-parent/",
			"http://jetpack-boost.test/parent/",
			"http://jetpack-boost.test/",
			"http://jetpack-boost.test/local-seo/",
			"http://jetpack-boost.test/test-page-2/"
		],
		"success_ratio": 0.5
	},
	{
		"key": "singular_my-post-type",
		"label": "Custom post type",
		"urls": ["http://jetpack-boost.test/my-post-type/this-is-a-post/"],
		"success_ratio": 0.5
	}
]
```

We can't get the core_front or core_posts pages via the rest API. There is a `/wp/v2/settings` endpoint from which we might be able to, but that requires authentication.

For success_ratio we can calculate based on the current logic we have in PHP. Everything else we have.

Going back to the endpoint approach. I'm doubting the privacy issue we raised initially. Is it an actual concern?

Everything we're listing/using is [public](https://github.com/Automattic/jetpack/blob/671fe1ee59da4b4d4152082612f47904b6605f00/projects/plugins/boost/app/lib/critical-css/source-providers/providers/Singular_Post_Provider.php#L130) and [visible](https://github.com/Automattic/jetpack/blob/671fe1ee59da4b4d4152082612f47904b6605f00/projects/plugins/boost/app/lib/critical-css/source-providers/providers/Singular_Post_Provider.php#L111) on the site (this applies for taxonomies as well).

I guess I got spooked for nothing in the beginning, but after reviewing the code, I think it's safe for us to leave this public. After all, all the data we use is already available on the website.

**Review by thingalon — APPROVED** [review]
I'm ok with this locked down to wpcom access only.

**Comment by dilirity** [comment]
**Update**

Doing it on WP.com is a no go. More info here p1710933796104259/1710855790.705649-slack-C016BBAFHHS

**Comment by dilirity** [comment]
@thingalon since we discussed not including cloud-css for the moment, should we close this?

**Comment by thingalon** [comment]
> @thingalon since we discussed not including cloud-css for the moment, should we close this?

sure :)

**Comment by dilirity** [comment]
Reopening this, as it's the only reliable way to fetch source providers from a website.

> I'm ok with this locked down to wpcom access only.

It will be locked down to wpcom access only.

**Review by pyronaur — APPROVED** [review]
:shipit:

**Pull request merged**